### PR TITLE
Arc CDI TCK related fixes, focused around behavior gated behind strict mode

### PR DESCRIPTION
**Pull request by manovotn**
Related to https://github.com/quarkusio/quarkus/issues/28558

This PR tackles tests that we earlier discover needed some `strictMode` behavior (in that Arc is too permissive compared to spec).
It also makes sure we can determine whether we are in strict compat. mode even in runtime.

I know @mkouba mentioned we could also introduce some abstraction for all parameters going into `Arc.initialize(...)` method. I can take a look at that tomorrow even though I am not wholly sold on doing that just yet because there are only two parameters and I am not aware of any more we need ATM. But like I said, I will still happily do that if requested :)

### CODE REVIEW

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:344]: Is this really necessary? I believe that we handle `Reception.IF_EXISTS` in the generated `ObserverMethod#notify()`...

File: independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java
Comment (mkouba) [inline independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java:18]: Is it guaranteed that `setInactive()`/`setActive()` is called from the same thread? I would at least make the field `volatile`.

**Review by mkouba — COMMENTED** [review]
>  I can take a look at that tomorrow even though I am not wholly sold on doing that just yet because there are only two parameters and I am not aware of any more we need ATM. But like I said, I will still happily do that if requested :)

Please do. There will be three `initialize()` methods... too many IMO and I'd hate to add more. I think that we should only keep one and deprecate the others.

File: independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java
Comment (manovotn) [reply independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java:18]: There is only a few usages in TCK (about 8?) some of which are CDI Full too and in none of them I saw multiple threads being used. But we can surely roll with `volatile` as well.
This is mostly a workaround for what I think should be a TCK challenge I have yet to create.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:344]: > Is this really necessary? I believe that we handle `Reception.IF_EXISTS` in the generated `ObserverMethod#notify()`...

It is, there are failures otherwise; note that this code is not here for reception handling.
Rather it is there to handle cases where an observer belongs to a bean with context that's inactive - so for instance req. context bean with observer but req. context wasn't activated. Originally, Arc was still trying to invoke it ending with `ContextNotActiveException`.
The dependent and `null` cases are for dependent beans which have no context of their own and for synth. observers which can belong to no bean.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:321]: > You could just modify the `EventImpl#createNotifier()` instead:

Not sure I understand?
The `boolean` present in `createNotifier()` now is used by init a shutdown events where we do not activate the context. Changing it the way you suggest would mean we'd start activating it for those events as well.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:344]: I see and I wonder if the spec acutally covers this use case. In any case, I would rather move the logic to the generated `ObserverMethod#notify()` so that it's all in one place.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:321]: Yes, I'm talking about the `createNotifier()` without the `activateRequestContext` param which is used for "normal" notifiers and has currently always `activateRequestContext=true`...

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:344]: > I see and I wonder if the spec acutally covers this use case

One of the TCKs I uncommented does exactly that - deactivate context, send notification and assert that it did not arrive.

> I would rather move the logic to the generated ObserverMethod#notify() so that it's all in one place.

Hm, to me it makes more sense outside of the method since this is effectively a condition based on which you don't want to notify an observer...

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:895]: The extra parameter is not necessary.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java:243]: Do we want to make this a part of public API?

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:344]: > > I see and I wonder if the spec acutally covers this use case
> 
> One of the TCKs I uncommented does exactly that - deactivate context, send notification and assert that it did not arrive.

Oook, it probably makes sense. Is it worth a clarification issue?

> 
> > I would rather move the logic to the generated ObserverMethod#notify() so that it's all in one place.
> 
> Hm, to me it makes more sense outside of the method since this is effectively a condition based on which you don't want to notify an observer...

Logically yes, but from maintainability POV it's better to change this [line](https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java#L417)  to `notify.ifNull(context).trueBranch().returnVoid();` :shrug:

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java:243]: The flag itself is part of the API since we use it to initialize the container. So I think that it's reasonable to add a read method as well.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:344]: This is already in [spec](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#observer_notification):

> If there is no context active for the scope to which the bean declaring the observer method belongs, then the observer method should not be called.

> Logically yes, but from maintainability POV it's better...

Fair enough.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:347]: I have a commit somewhere that adds a context object for the `@Dependent` pseudo-scope, which is always active. We'll be able to remove `!declaringBean.getScope().equals(Dependent.class)` when that is merged, if I understand this correctly.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java:243]: I originally didn't have it exposed but ended up doing that basically with the same reasoning Martin has here.
Not to mention it's nice not having to cast the result of `Arc.container()` to the impl class :)

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:347]: Yes, that is correct - except I will probably move this logic into the generated notify code.
I can add a comment about that.

Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:347]: > I have a commit somewhere that adds a context object for the @Dependent pseudo-scope, which is always active.

Hmm, I can't say I'm a fan... why is this needed? `@Dependent` is not a context anyway.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java
Comment (Ladicek) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java:243]: There's `ArcContainerImpl.instance()` to get the internal type, so that argument doesn't count IMHO :-) But if we need to expose that as a parameter to `Arc.initialize()`, then that makes sense. (For the record, I don't think we do, we could make it part of `ComponentsProvider` or something like that, so that we actually always propagate the build-time status to runtime instead of relying on the integrator to do it correctly. Of course that's a bit more complex.)

Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ArcContainer.java:243]: > (For the record, I don't think we do, we could make it part of ComponentsProvider or something like that, so that we actually always propagate the build-time status to runtime instead of relying on the integrator to do it correctly. Of course that's a bit more complex.)

So this is actually what I started looking into but after chat with Martin, I landed on current solution.

File: independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java
Comment (manovotn) [reply independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java:18]: It is very crude indeed :-D
Since I aimed to raise a TCK challenge to begin with, I just went for simplest possible fix. Not to mention, for something like CDI Lite, you can only really toggle on/off req. context and the class already work with `ManagedContext` which will only work for req. context in Arc.
Nonetheless, you are right to point that out; let's do the map approach instead.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:347]: Actually, @Ladicek you won't need to remove anything, moving this logic into observer generator doesn't require special case handling as dependent beans always have new instances attached anyway.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Arc.java
Comment (manovotn) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Arc.java:21]: I kept the no-args variant without deprecation as I think it's better than always requiring config object and passing in the default instance.

**Comment by manovotn** [comment]
Ok, round two @mkouba @Ladicek :)

Added two commits, first one addressed comments from this PR, second one introduces `ArcInitConfig` object along with a builder and replaces existing usages.

**Comment by manovotn** [comment]
CI failure is unrelated

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Arc.java
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Arc.java:32]: Pls add `forRemoval=true`...

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java:173]: We don't want to add more `TODO`s ;-). It's a package-private method so it shouldn't be hard to find out...

Comment (manovotn) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/EventImpl.java]: Ah, remnant of a note to myself :D
Removed

**Review by mkouba — APPROVED** [review]

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:1029]: Nitpicking: this comment clearly refers to the previous line (checking if the class is abstract or "concrete"). The line is moved pretty far away and it's no longer exactly clear what "add only concrete classes" means.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:1040]: Same here.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:1003]: Shouldn't we do a similar change for producer fields?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java
Comment (Ladicek) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java:417]: I'm probably just being stupid here, but conditional observers (`Reception.IF_EXISTS`) are handled above (in the exact same way), so this case only applies to "regular" obsevers (`Reception.ALWAYS`). Why do we need this change? Not saying we don't, just want to understand.

Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java:417]: Okay, I see that this probably fixes `ObserverNotificationTest.testObserverMethodNotInvokedIfNoActiveContext`.

**Review by Ladicek — APPROVED** [review]
Few minor comments and one bigger (we should probably treat producer fields similarly to producer methods).

I still think it would be better to pass build-time config to runtime through generated bytecode instead of relying on the integrator to setup the container correctly, but I agree that generating bytecode is generally a pain. And integrators should be sensible :-)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:1003]: Good catch!

File: independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java
Comment (mkouba) [reply independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java:21]: I agree with both comments.

Comment (mkouba) [reply independent-projects/arc/tcks/cdi-tck-porting-pkg/src/main/java/io/quarkus/arc/tck/porting/ContextsImpl.java:21]: Although it's still a fragile implementation of underspecified SPI... ;-)

**Comment by manovotn** [comment]
> Ah and sorry for late review, I need to attend to too many things these days :-)

No problem, I can adjust this PR but prob on Tue due to my PTO and public holidays.
Let me remove the waiting for CI label.

**Review by Ladicek — CHANGES_REQUESTED** [review]
Changing my review status to _request changes_ to signify that we don't want to merge this yet.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:1003]: This is now done for producer fields as well

**Comment by manovotn** [comment]
@Ladicek updated with requested changes (and rebased onto main), please re-review

**Review by Ladicek — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
